src/libs/Navigation/AppNavigator/createCustomRootStackNavigator.js

```diff
@@ -37,50 +39,74 @@ const propTypes = {
- * a screen, otherwise any child navigators won't be connected to the navigation tree properly.
+ * Returns the current descriptor key for the focused screen in this navigators state. The descriptor has a function
+ * called render() that we must call each time this navigator updates.
  *
  * @param {Object} props
  * @returns {Object}
  */
-function getCurrentViewDescriptor(props) {
+function getCurrentRouteDescriptorKey(props) {
     const currentRoute = props.state.routes[props.state.index];
-    const currentRouteKey = currentRoute.key;
-    const currentDescriptor = props.descriptors[currentRouteKey];
-    return currentDescriptor;
+    return currentRoute.key;
 }
 
-const ResponsiveView = props => (
-    <Modal
-        isVisible={props.currentURL
-            && props.currentURL.includes(props.path)}
-        backgroundColor={themeColors.componentBG}
-        type={CONST.MODAL.MODAL_TYPE.RIGHT_DOCKED}
-        onClose={Navigation.dismissModal}
-    >
-        {getCurrentViewDescriptor(props).render()}
-    </Modal>
-);
+const ResponsiveView = (props) => {
+    // Every descriptor in the navigation tree MUST render otherwise navigation tree gets screwed up.
+    // Modals on larger screens or web must always be rendered except we only really need the current view to display
+    // inside the modal not any views that are underneath the top Modal view. This works differently when we are on
+    // iOS and Android with small screen width since we will use the react-navigation modals instead.
+    const mainRoute = _.find(props.descriptors, descriptor => descriptor.options.isMainRoute);
+    const modalRoutes = _.filter(props.descriptors, (descriptor, key) => (
+
+        // We want all of the modal routes as long as it's not the current modal route since this should be rendered
+        // inside the modal itself.
+        descriptor.options.isModalRoute
+            && getCurrentRouteDescriptorKey(props) !== key
+    ));
+    const currentModalRoute = _.find(props.descriptors, (descriptor, key) => (
+        descriptor.options.isModalRoute
+            && getCurrentRouteDescriptorKey(props) === key
+    ));
+
+    const renderedModalScreen = currentModalRoute && currentModalRoute.render();
+    return (
+        <>
+            {mainRoute.render()}
+            <View style={styles.dNone}>
+                {_.map(modalRoutes, modalRoute => modalRoute.render())}
+                {renderedModalScreen}
+            </View>
+            {_.map(props.modalPaths, modalPath => (
+                <Modal
+                    key={`modal_${modalPath}`}
+                    isVisible={props.currentURL.slice(1).startsWith(modalPath)}
+                    backgroundColor={themeColors.componentBG}
+                    type={CONST.MODAL.MODAL_TYPE.RIGHT_DOCKED}
+                    onClose={Navigation.dismissModal}
+                >
+                    {renderedModalScreen || <View />}
```

Comment:
Trying to figure out some way around this double rendering. The challenge is react-nav must call `render()` on every descriptor each time the stack nav renders. But once we navigate out of a modal the view unmounts so it can't be displayed in the modal anymore.

One idea I had was to just list out all the modals naively to start. But this is pretty verbose and doesn't quite solve the issue of modal sub views.

There's probably a better way to handle this I just can't think of what it is.

Reply:
Solved this